ceno_zkvm/src/instructions/riscv/mulh.rs

```diff
@@ -7,195 +87,170 @@ use crate::{
+            constants::{BIT_WIDTH, UInt},
             r_insn::RInstructionConfig,
         },
     },
     uint::Value,
+    utils::i64_to_base,
     witness::LkMultiplicity,
 };
 use core::mem::MaybeUninit;
 
-/// This config handles R-Instructions that represent registers values as 2 * u16.
-#[derive(Debug)]
-pub struct ArithConfig<E: ExtensionField> {
-    r_insn: RInstructionConfig<E>,
+pub struct MulhInstructionBase<E, I>(PhantomData<(E, I)>);
 
-    rs1_read: UInt<E>,
-    rs2_read: UInt<E>,
-    rd_written: UIntMul<E>,
+pub struct MulhOp;
+impl RIVInstruction for MulhOp {
+    const INST_KIND: InsnKind = InsnKind::MULH;
 }
-
-pub struct MulhInstructionBase<E, I>(PhantomData<(E, I)>);
+pub type MulhInstruction<E> = MulhInstructionBase<E, MulhOp>;
 
 pub struct MulhuOp;
 impl RIVInstruction for MulhuOp {
     const INST_KIND: InsnKind = InsnKind::MULHU;
 }
 pub type MulhuInstruction<E> = MulhInstructionBase<E, MulhuOp>;
 
+pub struct MulhsuOp;
+impl RIVInstruction for MulhsuOp {
+    const INST_KIND: InsnKind = InsnKind::MULHSU;
+}
+pub type MulhsuInstruction<E> = MulhInstructionBase<E, MulhsuOp>;
+
+pub struct MulhConfig<E: ExtensionField> {
+    rs1_read: UInt<E>,
+    rs2_read: UInt<E>,
+    rd_written: UInt<E>,
+    sign_deps: MulhSignDependencies<E>,
+    prod_low: UInt<E>,
+    r_insn: RInstructionConfig<E>,
+}
+
+enum MulhSignDependencies<E: ExtensionField> {
+    UU {
+        constrain_rd: IsEqualConfig,
+    },
+    SU {
+        rs1_signed: Signed<E>,
+        rd_signed: Signed<E>,
+        constrain_rd: IsEqualConfig,
+    },
+    SS {
+        rs1_signed: Signed<E>,
+        rs2_signed: Signed<E>,
+        rd_signed: Signed<E>,
+    },
+}
+
 impl<E: ExtensionField, I: RIVInstruction> Instruction<E> for MulhInstructionBase<E, I> {
-    type InstructionConfig = ArithConfig<E>;
+    type InstructionConfig = MulhConfig<E>;
 
     fn name() -> String {
         format!("{:?}", I::INST_KIND)
     }
-
     fn construct_circuit(
         circuit_builder: &mut CircuitBuilder<E>,
-    ) -> Result<Self::InstructionConfig, ZKVMError> {
-        let (rs1_read, rs2_read, rd_written, rd_written_reg_expr) = match I::INST_KIND {
-            InsnKind::MULHU => {
-                // rs1_read * rs2_read = rd_written
-                let mut rs1_read = UInt::new_unchecked(|| "rs1_read", circuit_builder)?;
-                let mut rs2_read = UInt::new_unchecked(|| "rs2_read", circuit_builder)?;
-                let rd_written: UIntMul<E> =
-                    rs1_read.mul(|| "rd_written", circuit_builder, &mut rs2_read, true)?;
-                let (_, rd_written_hi) = rd_written.as_lo_hi()?;
-                (
-                    rs1_read,
-                    rs2_read,
-                    rd_written,
-                    rd_written_hi.register_expr(),
-                )
-            }
-
-            _ => unreachable!("Unsupported instruction kind"),
-        };
+    ) -> Result<MulhConfig<E>, ZKVMError> {
```

Comment:
Thanks for the documentation! it's great help to grasp the high level idea!

Giving currently implementation are strong coupling with field size, due to 64 bit representation got some overlap with Goldilock max field.
How about we also add some assertion, e.g. 

```
assert_eq!(UInt::<E>::TOTAL_BITS, u32::BITS as usize);
assert_eq!(E::BaseField::MODULUS_U64, goldilocks::MODULUS);
```
so once in the future we are having the different field implementation, then it can help to hint for implementation/constrain need to tweak a bit

Reply:
Yes, definitely agree!  Just added these to the start of `construct_circuit`.